### ISSUE TITLE
Recent Kubernetes tarballs from GitHub don't build using make (wrong linker args?)

### ISSUE THREAD

**Issue by mati865** [kubernetes/kubernetes#56346]
/kind bug

**What happened**:
```
+++ [1124 14:16:45] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
    k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
# k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
usage: link [options] main.o
(list of possible link args)
# k8s.io/kubernetes/hack/cmd/teststale
usage: link [options] main.o
(list of possible link args)
!!! [1124 14:16:46] Call tree:
!!! [1124 14:16:46]  1: /home/builder/aports/community/kubernetes/src/kubernetes-1.8.4/hack/lib/golang.sh:683 kube::golang::build_kube_toolchain(...)
!!! [1124 14:16:46]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1124 14:16:46] Call tree:
!!! [1124 14:16:46]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [Makefile.generated_files:322: _output/local/go/bin/deepcopy-gen] Error 1
make: *** [Makefile:512: generated_files] Error 2
```

**What you expected to happen**:
Working build.

**How to reproduce it (as minimally and precisely as possible)**:
Run `make -j1` in Kubernetes 1.8.4 source unpacked from tarball.

**Anything else we need to know?**:
Kubernetes 1.8.3 tarball and git v1.8.4 tag build fine.
Link to tarball: https://github.com/kubernetes/kubernetes/archive/v1.8.4.tar.gz

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.4
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): Alpine 3.6 and Ubuntu 17.10
- Kernel (e.g. `uname -a`): `Linux ubuntu-artful 4.13.0-16-generic #19-Ubuntu SMP Wed Oct 11 18:35:14 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: Alpine: go 1.8.4; Ubuntu: go 1.8.3, 1.9.1
- Others:

**Comment by mati865** [comment]
Tarball from https://dl.k8s.io/v1.8.4/kubernetes-src.tar.gz seems to build fine so the tarball v1.8.4 from GitHub must be mispackaged (downloaded it few times to make sure).

**Comment by akosiaris** [comment]
I just stumbled across this as well. And I can reliably reproduce it

`curl -L -O https://github.com/kubernetes/kubernetes/archive/v1.7.10.tar.gz`, try to build, works.

`curl -L -O https://github.com/kubernetes/kubernetes/archive/v1.7.11.tar.gz`, try to build, fails.

**Comment by schu** [comment]
I see the same issue with `1.8.5`.

**Comment by anguslees** [comment]
The bug is in the expansion of `-X foo/bar.gitTreeState=git archive` (note the space in the value).

`kube::version::ldflags` buiilds an array of words, then squashes them back to a single string, losing the original word boundaries.  This works for all values except KUBE_GIT_TREE_STATE="git archive", which is set in `kube::version::get_version_vars` for git archive exports, and obviously contains a space.

Fixed by #56249 (in 1.9.0), so only *all the 1.8 release tarballs from 1.8.4 onwards fail to build*.  Phew, that could have been serious.  :(

**Comment by ixdy** [comment]
I think I meant to cherrypick #56249 to 1.8 but didn't for some reason... 

We need better routing on issues, since I didn't notice this until just now. :(

cc @david-mcmahon

**Comment by ixdy** [comment]
This should be fixed in v1.8.7. Sorry again for missing this.
